weather_mv/loader_pipeline/regrid.py

```diff
@@ -32,16 +36,148 @@
+def _metview_op() -> t.Iterator[None]:
+    """Perform operation with MetView, including error handling and cleanup."""
+    try:
+        yield
+    except (ModuleNotFoundError, ImportError, FileNotFoundError) as e:
+        raise ImportError('Please install MetView with Anaconda:\n'
+                          '`conda install metview-batch -c conda-forge`') from e
+    finally:
+        _clear_metview()
+
+
+class MapChunkAsFieldset(beam.PTransform):
+    """Apply an operation with MetView on a xarray.Dataset as if it's a metview.Fieldset.
+
+    This transform will handle converting to and from xr.Datasets to mv.Fieldsets. This
+    allows the user to perform any MetView or Fieldset operation within the overridable
+    `apply()` method.
+
+    > Warning: This cannot process large Datasets without a decent amount of disk space!
+    """
+
+    def apply(self, key: xbeam.Key, fs: Fieldset) -> t.Tuple[xbeam.Key, Fieldset]:
+        return key, fs
+
+    def _apply(self, key: xbeam.Key, ds: xr.Dataset) -> t.Tuple[xbeam.Key, xr.Dataset]:
+        # Clear metadata so ecCodes doesn't mess up the conversion. Instead of default grib fields,
+        # ecCodes will use the parameter ID. Thus, the fields will appear in the final, regridded
+        # dataset.
+        for dv in ds.data_vars:
+            for to_del in ['GRIB_cfName', 'GRIB_shortName', 'GRIB_cfVarName']:
+                if to_del in ds[dv].attrs:
+                    del ds[dv].attrs[to_del]
+
+        with _metview_op():
+            # mv.dataset_to_fieldset() will error on input where there is only 1 value
+            # in a dimension. ECMWF's cfgrib is in its alpha version.
+            try:
+                fs = mv.dataset_to_fieldset(ds)
+            except ValueError as e:
+                raise ValueError(
+                    'please change `zarr_input_chunk`s so that there are no'
+                    'single element dimensions (e.g. {"time": 1} is not allowed).'
+                ) from e
+
+            # Apply any & all MetView or FieldSet operations.
+            kout, fs_out = self.apply(key, fs)
+
+            return kout, fs_out.to_dataset().compute()
+
+    def expand(self, pcoll):
+        return pcoll | beam.MapTuple(self._apply)
+
+
+@dataclasses.dataclass
+class RegridChunk(MapChunkAsFieldset):
+    """Regrid a xarray.Dataset with MetView.
+
+    Attributes:
+        regrid_kwargs: A dictionary of keyword-args to be passed into `mv.regrid()`
+            (excluding the dataset).
+        zarr_input_chunks: (Optional) When regridding Zarr data, how the input
+            dataset should be chunked upon open.
+    """
+    regrid_kwargs: t.Dict
+    zarr_input_chunks: t.Optional[t.Dict] = None
+
+    def template(self, source_ds: xr.Dataset) -> xr.Dataset:
+        """Calculate the output Zarr template by regridding a tiny slice of the input dataset."""
+        # Silence Dask warning...
+        with dask.config.set(**{'array.slicing.split_large_chunks': False}):
```

Comment:
This seems like a weird way to pass the arguments. Why not use regular keyword args?

Reply:
Will this key work that way if it is `.` delimited?

Reply:
Oh, I didn't catch that bit of weirdness. You're probably right that it has to be this way. Ugh, definitely not a fan of that.